src/swarm/neo/util/VoidBufferAsArrayOf.d

```diff
@@ -0,0 +1,93 @@
+/*******************************************************************************
+
+    Struct template which wraps a void[] with an API allowing it to be safely
+    used as an array of another type.
+
+    It is, of course, possible to simply cast a void[] to another type of array
+    and use it directly. However, care must be taken when casting to an array
+    type with a different element size. Experience has shown that this is likely
+    to hit undefined behaviour. For example, casting the array then sizing it
+    has been observed to cause segfaults, e.g.:
+
+    ---
+        void[]* void_array; // acquired from somewhere
+
+        struct S { int i; hash_t h; }
+        auto s_array = cast(S[]*)void_array;
+        s_array.length = 23;
+    ---
+
+    The exact reason for the segfaults is not known, but this usage appears to
+    lead to corruption of internal GC data (possibly type metadata associated
+    with the array's pointer).
+
+    Sizing the array first, then casting is fine, e.g.:
+
+    ---
+        void[]* void_array; // acquired from somewhere
+
+        struct S { int i; hash_t h; }
+        (*void_array).length = 23 * S.sizeof;
+        auto s_array = cast(S[])*void_array;
+    ---
+
+    The helper VoidBufferAsArrayOf simplifies this procedure and removes the
+    risk of undefined behaviour by always handling the void[] as a void[]
+    internally.
+
+    Copyright: Copyright (c) 2017 sociomantic labs GmbH. All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module swarm.neo.util.VoidBufferAsArrayOf;
+
+import ocean.transition;
+
+/*******************************************************************************
+
+    Struct template which wraps a void[] with an API allowing it to be safely
+    used as an array of another type.
+
+    Params:
+        T = element type of array which the API mimics
+
+*******************************************************************************/
+
+public struct VoidBufferAsArrayOf ( T )
+{
+    static assert(!is(T == void));
+
+    private void[]* buffer;
+
+    invariant ( )
+    {
+        assert(this.buffer.length % T.sizeof == 0);
+    }
+
+    public size_t length ( )
+    {
+        return this.buffer.length / T.sizeof;
+    }
+
+    public void length ( size_t len )
+    {
+        this.buffer.length = len * T.sizeof;
+        enableStomping(*this.buffer);
+    }
+
+    public void opCatAssign ( T elem )
+    {
+        this.length = this.length + 1;
+        auto elem_in_array =
+            cast(T*)this.buffer.ptr + ((this.length - 1) * T.sizeof);
+        *elem_in_array = elem;
```

Comment:
You can just append to `(*this.buffer)`, and there should also be an `opCatAssign` overload to append an array:
```d
T[] opCatAssign ( in T[] arr )
{
    return cast(T[])((*this.buffer) ~= arr);
}
```
`arr` is automatically cast to `void[]`. (The real `~=` operator returns the array.) With this overload appending a single element is as simple as
```d
T[] opCatAssign ( in T element )
{
    return this.opCatAssign((&element)[0 .. 1]);
}
```